### ISSUE TITLE
Question:  sp/sharing:  getSharingInformation or SharedWithDetails

### ISSUE THREAD

**Issue by mikezimm** [pnp/pnpjs#1695]
### Category
- [ ] Enhancement
- [ ] Bug
- [x] Question
- [ ] Documentation gap/issue

### Version
Latest

### Expected / Desired Behavior / Question
I want to get details on all items in a library that have been shared and am looking for the easiest way...

I noticed the item return properties has this field.
![image](https://user-images.githubusercontent.com/49648086/114290717-95b9ae80-9a4f-11eb-9795-c099538c3ee3.png)

I want to just find all the items with that property and get the details.

I'm guessing I could use https://pnp.github.io/pnpjs/sp/sharing/#getsharinginformation but for some reason I get a 404 error.
```typescript
const info = await sp.web.getFolderByServerRelativeUrl("/sites/MySiteURL/Shared%20Documents/").getSharingInformation();
```
When I run that line of code, I get 404 error and I ahve double checked my Urls... am I missing something else?

### Observed Behavior
![image](https://user-images.githubusercontent.com/49648086/114290732-ac600580-9a4f-11eb-8852-60e850dac486.png)

**Comment by mikezimm** [comment]
I noticed that on the item level there is also the SharedWithUsersId property which I can filter on using:
```typescript
const sharedItems = await thisListObject.items.filter('SharedWithUsersId ne null').get();
```

The unfortunate limitation that I can't get the filename which makes it kind of useless as is.

If I go the route of returning files, then I get the file names but no SharedWith properties.  Plus I think I would have to check every folder which could be kind of complicated.

Can I do something with an SP search query that will get me all the files with SharedWithUsersId in ALL the folders of a library and also get the filename/url?

Thanks!

**Comment by mikezimm** [comment]
I tried this in a search query and was able to get some information
**but nothing that I could use to get a clickable path to the file**... just the name...

Is there another managed property I can get?
Or a way to use the DocId or whatever the UniqueId is to get a link or open the file?

Thanks again!

```typescript
    let query=`Path:https://tenant.sharepoint.com/sites/PivotNotInstalled/Shared%20Documents* AND ContentClass:STS_ListItem` ;
    let shareSelect = ["*","Title", "ServerRelativeUrl", "ServerRelativePath", "ID", "Id", "Path", "Filename","FileLeafRef", "Author","Editor","SharedWithUsersId", "SharedWithDetails", "SharedWithInternal","FileLeafRef","ows_FileLeafRef","ModifiedBy","Created","Modified","CreatedBy","CreatedById","ModifiedById","ServerRedirectedEmbedURL","ServerRedirectedPreviewURL","ServerRedirectedPreviewURL"];
```
Here's a screenshot of the return results:
![image](https://user-images.githubusercontent.com/49648086/114294020-251f8b80-9a69-11eb-95a5-24c17b3c861d.png)

**Comment by patrick-rodgers** [comment]
If I try your original request it works for me as expected:

```
// works
const r = await sp.web.getFolderByServerRelativeUrl("/sites/dev/Shared%20Documents/test-folder96").getSharingInformation();
```
I did notice your url in the screen shot appears to show you getting sharing information for a web? Web does indeed not support use of "getSharingInformation". A little confused by that as the code and screen shots don't seem to match up.

I am not entirely sure what you are after as the thread is a bit hard to follow. IF your goal is to get all the items shared with a given user the [graph's insights api](https://docs.microsoft.com/en-us/graph/api/insights-list-shared?view=graph-rest-1.0) might help. We don't have the endpoints coded for that currently but we can. IF your goal is to locate all the things that have sharing enabled without traversing all the folders/items I am not aware of an API to accomplish that.

Perhaps you can summarize what you are trying to accomplish as we can provide some more guidance?

**Comment by mikezimm** [comment]
@patrick-rodgers ,
In the end, my goal was to get all sharing details on all items in all folders of a library without having to loop through all the nested folders.

I made that qualification because I presume sp.web.getFolderByServerRelativeUrl("/sites/dev/Shared%20Documents/test-folder96") only gets items in the test-folder96 and not all it's sub-folders correct?

## Anyway, I was able to get the information now using this method:
basically get all of the "items" on the library, add FileRef and FileLeafRef to the select.
Then just loop through the results and manually extract a list of items that were shared.
```typescript
sp.web.lists.getByTitle('Documents').items.select("Title","FileRef","FileLeafRef","SharedWithUsers/Title","SharedWithUsers/Name","SharedWithDetails").expand("SharedWithUsers")
```

As you can see by the screenshot, I do get the sharing details along with the Filename and Url.
I think my next step will be to see if there is still a way to add a filter to the rest call so that I don't have to get everything and then search for items I want.  But this is progress and I can live with it.

Since I was able to get this to work, I'll close the question.  I will also have to go back and see if I can get the .getSharingInformation() to actually work... not sure what I'm doing wrong but I'll figure it out.

Thanks again for the reply

![image](https://user-images.githubusercontent.com/49648086/116573854-2a8b2b80-a8db-11eb-93ce-a6ab1d9a6a20.png)
